Assembler.UnitTests/AssemblerTests.cs

```diff
@@ -0,0 +1,165 @@
+        private Base.MessagesAssembler _assembler;
+
+        private List<BaseAssembledMessage> _assembledMessages;
+        private List<Mock<IHandler<BaseFrame, BaseMessageInAssembly>>> _handlersList;
+
+        [SetUp]
+        public void Setup()
+        {
+            _assembledMessages = new List<BaseAssembledMessage>();
+            _resolverMock = new Mock<IResolver<FrameType, IHandler<BaseFrame, BaseMessageInAssembly>>>();
+            _converterMock = new Mock<IConverter<BaseMessageInAssembly, BaseAssembledMessage>>();
+            _cacheMock = new Mock<ITimeBasedCache<BaseMessageInAssembly>>();
+
+            _firstHandlerMock = new Mock<IHandler<BaseFrame, BaseMessageInAssembly>>();
+            _secondHandlerMock = new Mock<IHandler<BaseFrame, BaseMessageInAssembly>>();
+
+            _handlersList = new List<Mock<IHandler<BaseFrame, BaseMessageInAssembly>>>
+            {
+                _firstHandlerMock,
+                _secondHandlerMock
+            };
+            var handlersList = _handlersList.Select(handlers => handlers.Object);
+
+            _assembler = new Base.MessagesAssembler(_resolverMock.Object, _cacheMock.Object, handlersList,
+                _converterMock.Object, Utilities.GetLoggerFactory());
+
+            _assembler.MessageAssembled += _assembledMessages.Add;
+        }
+
+        [TearDown]
+        public void Teardown()
+        {
+            _resolverMock.VerifyNoOtherCalls();
+            _converterMock.VerifyNoOtherCalls();
+            _cacheMock.VerifyNoOtherCalls();
+            _firstHandlerMock.VerifyNoOtherCalls();
+            _secondHandlerMock.VerifyNoOtherCalls();
+        }
+
+        [Test]
+        public void Constructor_NoHandlers_NoExceptionBeingThrown()
+        {
+            // Arrange
+            var emptyHandlersList = new List<IHandler<BaseFrame, BaseMessageInAssembly>>();
+
+            // Act + Assert
+            Assert.DoesNotThrow(() => _assembler = new Base.MessagesAssembler(_resolverMock.Object, _cacheMock.Object, emptyHandlersList,
+                _converterMock.Object, Utilities.GetLoggerFactory()));
+            Assert.Zero(_assembledMessages.Count);
+        }
+
+        [TestCase(0)]
+        [TestCase(1)]
+        public void ReleaseMessage_FirstHandlerRaisesEvent_AssemblerConvertsAndRaises(int handlerIndex)
+        {
+            // Arrange
+            var messageInAssembly = new Mock<BaseMessageInAssembly>();
+            var assembledMessage = new Mock<BaseAssembledMessage>();
+
+            _converterMock.Setup(converter => converter.Convert(It.IsAny<BaseMessageInAssembly>()))
+                .Returns(assembledMessage.Object);
+
+            // Act
+            _handlersList[handlerIndex].Raise(handler => handler.MessageAssemblyFinished += null, messageInAssembly.Object);
+
+            // Assert
+            _converterMock.Verify(converter => converter.Convert(It.IsAny<BaseMessageInAssembly>()), Times.Once);
```

Comment:
If you have the line after this one, wouldn't `VerifyNoOtherCalls()` already assert that?

Reply:
If so, notice that you have this check in a bunch of other places as well (otherwise, no need to change these)

Reply:
Sorry, I don't think I get you.
`VerifyNoOtherCalls` verifies that nothing being called that you haven't set up.

Reply:
You verify `converter.Convert(messageInAssembly.Object)` because you want to know that this method was called with the right object, but you also verify `converter.Convert(It.IsAny<BaseMessageInAssembly>())` because you want to know that it was called with any instance of `BaseMessageInAssembly` just once (**I assume**).

What I ask is - if you verify `converter.Convert(messageInAssembly.Object)` then call `_converterMock.VerifyNoOtherCalls()` - wouldn't that be the same?

Reply:
Ok I got it.
It isn't the same because when I set it up I set it up to `ItIsAny`, so the `VerifyNoOtherCalls()` wouldn't return anything.
I think we can set it up for a specific instance and then we won't have to double check everytime.
@omerkurtz What do you say?